text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,266 @@
+
+<figure>
+  <img src="../assets/0000-ssr-rehydration-fig2.png" alt="Fig. 2" width="60%">
+  <figcaption align = "center">
+    <b>Fig.2: the proposed implementation</b>
+  </figcaption>
+</figure><br /><br />
+
+In roughly chronological order with respect to a single page request, the following steps shall be taken as part of rehydration:
+
+- As part of SSR, serialize render-critical state and bake it into the HTML response returned from the server. For each `LightningElement`, attach corresponding state to its root DOM node as `data-ssr-state`.
+
+- Rather than calling `LWC#createElement`, required that `LWC#hydrateElement` is called to mount the root Lightning Web Component on the client.
+
+- As the initial client-side render progresses, extract state from `data-ssr-state` attributes (which can be attached at any LWC root node), instantiate `LightningElements` with extracted state, and generate client-side VDOM.
+
+- Patch server-side VDOM with client-side VDOM.
+  
+  - Where `data-ssr-preserve` is detected, extract the server-side VDOM subtree and store for subsequent client-side renders.
+  
+  - Attach event listeners and non-serializable props.
+  
+  - Resolve unexpected inconsistencies in favor of the client-side VDOM, addressing [Exception C to the happy path](#the-happy-path). With the exception of placeholder components, this case should be rare, since the first CSR will rely on state baked into SSR as `data-ssr-state`.
+
+### Render-critial state
+
+For the purposes of SSR rehydration, render-critical state is defined as being:
+
+- owned by a Lightning Web Component
+- decorated by `@api` or `@track`
+- serializable into JSON
+- able to be generated synchronously during SSR
+
+State that is not serializable to JSON but is decorated with `@api` or `@track` will be unsupported for SSR rehydration. This constraint should be clearly documented. We should also consider surfacing a warning when not in production. **Note:** If this restriction proves too constraining, we can explore providing mechanisms for consumers to marshal their non-serializable state into a serializable form and back again.
+
+State that is contained outside of a LWC (e.g. in a store of some kind) will need to be serialized manually during SSR and rehydrated prior to invocation of `hydrateElement` on the client.
+
+### Bailing out
+
+There is a spectrum of choices possible when encountering an unexpected mismatch during rehydration. The following is a sampling from that spectrum:
+
+1. Bail out of rehydration entirely when a mismatch is detected, throw away the original HTML, and force a full re-render.
+2. Handle easy mismatches (text nodes, certain attrs, etc.) falling back to number 1 for anything more extreme.
+3. Bail out of rehydration for subtrees, rooted at a mismatch.
+4. Bail out of rehydration for subtrees, rooted at the component where the mismatch occurred.

```

Comment:
Later on you say that the way a user can opt-in/out-out to have a different template on Client vs Server is in user land by swapping the whole template. I think it will be interesting to also put an example where the granularity is at the statement level. (ex. `<template if:true={inServer}>FOO<else>BAR`). How how that would affect the bailing out decision.

Reply:
I've clarified that this is an example, only intended to show what is possible, not to prescribe a particular implementation.